app/src/ui/diff/diff-helpers.tsx

```diff
@@ -454,3 +454,39 @@ export function getFirstAndLastClassesUnified(
 
   return classNames
 }
+
+/**
+ * Compares two text diffs for structural equality.
+ *
+ * Components needing to know whether a re-render is necessary after receiving
+ * a diff is the intended use case.
+ */
+export function textDiffEquals(
+  x: ITextDiff | ILargeTextDiff,
+  y: ITextDiff | ILargeTextDiff
+) {
+  if (x === y) {
+    return true
+  }
+
+  if (
+    x.text === y.text &&
+    x.kind === y.kind &&
+    x.hasHiddenBidiChars === y.hasHiddenBidiChars &&
+    x.lineEndingsChange === y.lineEndingsChange &&
+    x.hunks.length === y.hunks.length
+  ) {
+    // This is a performance optimization which lets us avoid iterating over all
+    // lines (deep equality on all hunks). We're already comparing the diff text
+    // above so the only thing that can change with the diff text staying the
+    // same is whether or not the last line is followed by a trailing newline.
+    // That information is encodeded in the noTrailingNewLine property which
+    // exists on all lines but is only ever set on lines in the last hunk
+    return (
+      x.hunks.length === 0 ||
+      x.hunks[x.hunks.length - 1].equals(y.hunks[y.hunks.length - 1])
+    )
```

Comment:
Given the whole diff could be one hunk, couldn't we just compare the last line of both diffs?

Reply:
That would likely cover almost all cases yes. At least for "normal" use but it is possible to have two diffs where the text is equal, neither (or both) has the `noTrailingNewLine` property set on the last line but yet they are not equal. Here's one such example:

```diff
diff --git a/README.md b/README.md
index bab18e6..ffcaf82 100644
--- a/README.md
+++ b/README.md
@@ -1,2 +1,3 @@
 Foo
-Bar
\ No newline at end of file
+BarBar
+Baz
```

That change removed the last line of the file and replaced it with `BarBar\nBaz\n`. If the previous file did have a trailing newline the diff would look like this and the text would be equal.

```diff
diff --git a/README.md b/README.md
index bab18e6..ffcaf82 100644
--- a/README.md
+++ b/README.md
@@ -1,2 +1,3 @@
 Foo
-Bar
+BarBar
+Baz
```

Would you ever encounter this in real world usage? I don't know, and given how this only affects presentation we would likely get away with it even if it did but my head hurt too much to reason about it that I figured it was just easier to take the safer but teensy bit slower approach.